### PR TITLE
Step 2: Scaffolding and notifications for "SFDX: Capture Startup Performance" command

### PR DESCRIPTION
**Pull request by daphne-sfdc**
- Wire the "SFDX: Capture Startup Performance" command so that it prints a message to the console when run
- Add notifications to indicate when the command is running, completed successfully, or failed with an error
- Add pseudocode and scaffolding functions for future expected behavior of the command

@W-16385161@

### CODE REVIEW

File: src/captureStartupPerformance.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      await copyContentsToFile(parsedResults);
+      const parsedResultsFile = await saveCurrentFile();
+      // send parsed results to appinsights
+      await sendToAppInsights(parsedResultsFile);
+
+      await sleep(5000);
+    });
+
+    // Show success notification to the user
+    vscode.window.showInformationMessage('SFDX: Capture Startup Performance command completed successfully.');
```
Comment (peternhale) [inline src/captureStartupPerformance.ts:48]: Why a separate notification for completion?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const registerCommands = (
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  extensionContext: vscode.ExtensionContext
+): vscode.Disposable => {
+  const captureStartupPerformanceCmd = vscode.commands.registerCommand(
+    'sf.capture.startup.performance',
+    captureStartupPerformance
+  );
+
+  return vscode.Disposable.from(captureStartupPerformanceCmd);
```
Comment (peternhale) [inline src/index.ts:19]: register command returns a disposable already and vscode.Disposable.from is used to "Combine many disposable-likes into one. You can use this method when having objects with a dispose function which aren't instances of Disposable."
```suggestion
  return vscode.commands.registerCommand(
    'sf.capture.startup.performance',
    captureStartupPerformance
  );
```

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/captureStartupPerformance.ts (hunk 1 above)
Comment (daphne-sfdc) [reply src/captureStartupPerformance.ts:47]: I'm making this consistent with all our other Salesforce extensions.  When our other SFDX commands complete successfully, we get a notification indicating that.

(Although, I am quite surprised that the notification is now replaced with a message in the status bar now.)
Screenshot 2024-08-21 at 10 47 58 AM

File: src/captureStartupPerformance.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      await openStartupPerformanceFile();
+      // save startup performance file (TODO: how to do this without popup dialog?)
+      const fileName = await saveCurrentFile();
+      // read startup performance file using fs.readFileSync()
+      const fileContents = await readStartupPerformanceFile(fileName);
+      // parse startup performance file using marked
+      await sleep(2000);
+      const parsedResults = await parseStartupPerformanceFile(fileContents);
+      // save parsed results to a file
+      await copyContentsToFile(parsedResults);
```
[the remark this answers is not in the document]
Comment (daphne-sfdc) [reply src/captureStartupPerformance.ts:38]: I was thinking that because they do two separate things, they would be separated.  I called `saveCurrentFile()` twice in the `captureStartupPerformance()` function.

The first time, it is called with the Startup Performance file open in Editor view.  I want to save that file, which can be saved directly into another file without having to explicitly copy the content into another file.

The second time, it is used to save the results of parsing the Startup Performance file.  I need to first call `copyContentsToFile()` to copy the parsed results into a separate new file, then call `saveCurrentFile()` on the new file with the parsed results in it.

File: src/captureStartupPerformance.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      progress.report({ message: 'Running...' });
+
+      // open Developer: Startup Performance (open via API - try "perfview.show")
+      await openStartupPerformanceFile();
+      // save startup performance file (TODO: how to do this without popup dialog?)
+      const fileName = await saveCurrentFile();
+      // read startup performance file using fs.readFileSync()
+      const fileContents = await readStartupPerformanceFile(fileName);
+      // parse startup performance file using marked
+      await sleep(2000);
```
Comment (peternhale) [inline src/captureStartupPerformance.ts:35]: Are the calls to sleep temorary

File: src/captureStartupPerformance.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+const parseStartupPerformanceFile = async(fileContents: string): Promise<string> => {
+  // throw new Error('Error thrown in parseStartupPerformanceFile()');
+  return 'parsed contents of Extension Activation Stats table';
+}
+
+const copyContentsToFile = async(contents: string): Promise<void> => {}
+
+const sendTelemetryData = async(fileName: string): Promise<void> => {}
+
+// NOTE: This function and the lodash dependency are used only for testing purposes.
```
Comment (peternhale) [inline src/captureStartupPerformance.ts:75]: update comment and I would suggest moving the function to a util, given that sleeping has nothng to do with the purpose of this file.

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as vscode from 'vscode';
+import { captureStartupPerformance } from './captureStartupPerformance';
+
+const registerCommands = (
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  extensionContext: vscode.ExtensionContext
```
Comment (peternhale) [inline src/index.ts:12]: If it is not going to be used, why have it as a parameter?

File: package.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  "main": "./out/src/index.js",
+  "dependencies": {},
+  "devDependencies": {
+    "@types/node": "^18.11.9",
+    "@types/jest": "^29.5.12",
+    "@types/vscode": "^1.86.0",
+    "@tsconfig/node18": "^18.2.4"
+  },
+  "scripts": {
+    "vscode:prepublish": "npm run compile && npm run bundle:extension && npm run update-for-packaging",
```
Comment (peternhale) [inline package.json:27]: Thank you

**Review by peternhale — CHANGES_REQUESTED** [review]
A few more nits

File: src/captureStartupPerformance.ts (hunk 4 above)
Comment (daphne-sfdc) [reply src/captureStartupPerformance.ts:35]: Yes, I added them for the purpose of testing the notifications.  I'll remove them once the real code is written.

**Review by peternhale — APPROVED** [review]

File: .prettierrc [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "printWidth": 80,
```
Comment (peternhale) [inline .prettierrc:2]: The notion print width of 80 columns is so old school.
[This is where it started](https://en.wikipedia.org/wiki/Punched_card)

File: images/Ruth_trapeze.png
Comment (peternhale) [inline images/Ruth_trapeze.png:1]: Should probably delete this file.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,91 @@
+{
+  "name": "salesforcedx-vscode-capture-startup-performance",
+  "displayName": "Capture Extension Startup Performance",
+  "description": "Extension that captures the Extension Activation Stats from 'Developer: Startup Performance' and reports the data to Azure AppInsights",
```
Comment (peternhale) [inline package.json:4]:
``` suggestion
  "description": "Extension that captures the Extension Activation Stats from 'Developer: Startup Performance' and reports the data via telemetry",
```

File: package.json [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  "version": "1.0.0",
+  "publisher": "salesforce",
+  "repository": {
+    "url": "https://github.com/forcedotcom/salesforcedx-vscode-capture-startup-performance"
+  },
+  "icon": "images/Ruth_trapeze.png",
```
Comment (peternhale) [inline package.json:10]: Let's see if we can find another image to use.

File: src/checkOpenSalesforceProject.ts
Comment (peternhale) [inline src/checkOpenSalesforceProject.ts:1]: Why is this needed? The pre-condition for the extension activating should be guarded by "needs to be in a dx project."

**Review by peternhale — CHANGES_REQUESTED** [review]

File: .prettierrc (hunk 8 above)
Comment (daphne-sfdc) [reply .prettierrc:2]: 😂 I got that line from the .prettierrc file in salesforcedx-vscode.  I'll remove it here but then we should also remove it there too.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      "LICENSE",
+      "package.nls.json",
+      "README.md",
+      "images",
+      "dist"
+    ],
+    "packageUpdates": {
+      "main": "dist/index.js",
+      "dependencies": {
+        "applicationinsights": "^3.2.2",
```
Comment (peternhale) [inline package.json:47]: @daphne-sfdc why are these in package updates section?

File: scripts/vsce-bundled-extension.ts [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+import { chdir } from 'process';
+import * as shell from 'shelljs';
+
+const logger = (msg: string, obj?: any) => {
+  if (!obj) {
+    console.log(`*** ${msg}`);
+  } else {
+    console.log(`*** ${msg}`, obj);
+  }
+};
```
Comment (peternhale) [inline scripts/vsce-bundled-extension.ts:18]:
``` suggestion
const logger = (msg: string, ...obj: any[]) => {
    console.log(`*** ${msg}`, ...obj);
};
```

**Review by peternhale — APPROVED** [review]
@daphne-sfdc just a couple of nits that can be picked up on your next PR.

File: package.json (hunk 11 above)
Comment (daphne-sfdc) [reply package.json:47]: I got that from the `packaging` section of package.json of salesforcedx-vscode-core.
```
    "packageUpdates": {
      "main": "dist/src/index.js",
      "dependencies": {
        "applicationinsights": "1.0.7",
        "shelljs": "0.8.5"
      },
      "devDependencies": {}
    }
```
